README.rst
Introduction
============

Core wrapper library for RDF handling in PMR2.


docs/HISTORY.rst
Changelog
=========

1.0 - Released (2016-03-08)
---------------------------

* Major version bump to modern version of rdflib.

0.4 - Released (2013-07-08)
---------------------------

* Added view that shows the extracted RDFXML.

0.3 - Released (2011-08-17)
---------------------------

* Added a simple helper method.

0.2 - Released (2011-04-08)
---------------------------

* Minor fixes related to packaging of this package.


0.1 - Released (2010-06-21)
---------------------------

* A very basic RDF abstraction layer for PMR2.
* Serves as a base for other RDF function on files within PMR2.



pmr2/rdf/tests/test_base.py
import unittest
import os
from os.path import dirname, join
from cStringIO import StringIO

import rdflib

from pmr2.rdf.base import RdfXmlObject
from pmr2.rdf.base import RdfXmlMetadata

testroot = dirname(__file__)
input_dir = join(testroot, 'input')
output_dir = join(testroot, 'output')

class BaseTestCase(unittest.TestCase):

    def setUp(self):
        self.file = RdfXmlObject()

    def tearDown(self):
        pass

    def parse(self, filename):
        fd = open(filename)
        try:
            self.file.parse(fd)
        finally:
            fd.close()

    def test_0000_basic(self):
        self.parse(join(input_dir, 'well-formed-3-node.xml'))
        self.assertEqual(len(self.file.subgraphIds), 3)

    def test_0001_simple(self):
        self.parse(join(input_dir, 'simple.rdf'))
        self.assertEqual(len(self.file.subgraphIds), 1)


class MetadataTestCase(unittest.TestCase):

    def setUp(self):
        self.file = RdfXmlMetadata()

    def tearDown(self):
        pass

    def parse(self, filename):
        fd = open(filename)
        try:
            self.file.parse(fd)
        finally:
            fd.close()

    def test_0000_queryEFNote(self):
        self.parse(join(input_dir, 'well-formed-3-node.xml'))
        answer = {
            u'key1': u'value1',
            u'key2': u'value2',
        }
        # must convert to dictionary for testign because of multiple 
        # results being unordered.
        results = dict(self.file.queryEFNote('ex'))
        self.assertEqual(answer, results)

    def test_0100_queryDC(self):
        self.parse(join(input_dir, 'well-formed-3-node.xml'))
        answer = [(
            ('dc:title', u'Test XML 3 node'),
            ('dc:creator', u'Tommy'), 
            ('dc:description', u'A test RDF/XML document'),
        )]
        # single result set, query ordered.
        results = self.file.queryDC('')
        self.assertEqual(answer, results)

    def test_0101_queryDC_partial(self):
        self.parse(join(input_dir, 'partial_dc.rdf'))
        answer = [(
            ('dc:title', u'Test RDF File'),
            ('dc:creator', u'Example'), 
            ('dc:description', None),
        )]
        # single result set, query ordered.
        results = self.file.queryDC('')
        self.assertEqual(answer, results)


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(BaseTestCase))
    suite.addTest(unittest.makeSuite(MetadataTestCase))
    return suite

if __name__ == '__main__':
    unittest.main()



pmr2/rdf/tests/test_browser.py
from Zope2.App import zcml
from Products.Five import fiveconfigure
from Products.PloneTestCase.layer import onsetup, onteardown

from pmr2.app.workspace.content import WorkspaceContainer, Workspace
from pmr2.app.workspace.tests import storage
from pmr2.app.workspace.tests import base

from pmr2.app.exposure.content import ExposureContainer, Exposure, ExposureFile

@onsetup
def setup():
    import pmr2.rdf
    fiveconfigure.debug_mode = True
    zcml.load_config('configure.zcml', pmr2.rdf)
    fiveconfigure.debug_mode = False

@onteardown
def teardown():
    pass

setup()
teardown()

storage.DummyStorageUtility._dummy_storage_data['rdftest'] = [
    {
    'test.rdf': """\
<?xml version="1.0" encoding="utf-8"?>
<ex xmlns="http://example.com/ex/1.0#"
    xmlns:dc="http://purl.org/dc/elements/1.1/"
    xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">

  <node>
    <node>
      <rdf:RDF>
        <rdf:Description rdf:about="#item">
          <dc:creator>Example User</dc:creator>
          <dc:title>Test RDF Document</dc:title>
        </rdf:Description>
      </rdf:RDF>
    </node>
  </node>

  <line>
    <rdf:RDF>
      <rdf:Description rdf:about="#item2">
        <dc:creator>Second User</dc:creator>
        <dc:title>Second Item</dc:title>
      </rdf:Description>
    </rdf:RDF>
  </line>

</ex>
""",

    'brokenxml.rdf': """\
<?xml version="1.0" encoding="utf-8"?>
<ex xmlns="http://example.com/ex/1.0#"
    xmlns:dc="http://purl.org/dc/elements/1.1/"
    xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">

  <node>
    <node>
      <rdf:RDF>
        <rdf:Description rdf:about="#item">
          <dc:creator>Example User</dc:creator>
          <dc:title>Test RDF Document</dc:title>
        </rdf:Description>
      </rdf:RDF>
  </node

</ex>
""",

}]

extracted = """\
<?xml version="1.0" encoding="UTF-8"?>
<rdf:RDF
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
>
  <rdf:Description rdf:about="#item">
    <dc:creator>Example User</dc:creator>
    <dc:title>Test RDF Document</dc:title>
  </rdf:Description>
  <rdf:Description rdf:about="#item2">
    <dc:creator>Second User</dc:creator>
    <dc:title>Second Item</dc:title>
  </rdf:Description>
</rdf:RDF>
"""

empty = """\
<?xml version="1.0" encoding="UTF-8"?>
<rdf:RDF
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
>
</rdf:RDF>
"""

class RdfViewDocTestCase(base.WorkspaceDocTestCase):

    def setUp(self):
        super(RdfViewDocTestCase, self).setUp()
        self.portal['workspace'] = WorkspaceContainer()
        w = Workspace('rdftest')
        w.storage = 'dummy_storage' 
        self.portal.workspace['rdftest'] = w

        self.portal['exposure'] = ExposureContainer()
        e = Exposure('rdftest')
        e.workspace = u'/plone/workspace/rdftest'
        e.commit_id = u'0'
        self.portal.exposure['rdftest'] = e
        ef = ExposureFile('test.rdf')
        self.portal.exposure.rdftest['test.rdf'] = ef

    def test_0000_base(self):
        self.testbrowser.open(self.portal.absolute_url() +
            '/workspace/rdftest/pmr2_rdf/0/test.rdf')
        contents = self.testbrowser.contents
        self.assertTrue('<dc:creator>Example User</dc:creator>' in contents)
        self.assertTrue('<dc:title>Test RDF Document</dc:title>' in contents)
        self.assertTrue('<dc:creator>Second User</dc:creator>' in contents)
        self.assertTrue('<dc:title>Second Item</dc:title>' in contents)

        self.assertEqual(self.testbrowser.headers['Content-type'],
            'application/rdf+xml')

    def test_0001_broken(self):
        self.testbrowser.open(self.portal.absolute_url() +
            '/workspace/rdftest/pmr2_rdf/0/brokenxml.rdf')
        contents = self.testbrowser.contents
        self.assertEqual(contents, empty)

    def test_0100_exposure_file_redirect(self):
        self.testbrowser.open(self.portal.absolute_url() +
            '/exposure/rdftest/test.rdf/pmr2_rdf')
        self.assertEqual(self.testbrowser.url,
            'http://nohost/plone/workspace/rdftest/@@pmr2_rdf/0/test.rdf')


pmr2/rdf/browser.py
from cStringIO import StringIO

import zope.interface
import zope.component

from pmr2.app.workspace.browser.browser import FilePage
from pmr2.app.exposure.browser.browser import ExposureFileRedirect

from pmr2.rdf.base import RdfXmlObject


class RdfPage(FilePage):

    def render(self):
        super(RdfPage, self).update()

        contents = self.data['contents']()
        s = StringIO(contents)
        rdf = RdfXmlObject()
        try:
            rdf.parse(s)
        except:
            pass
        contents = rdf.graph.serialize()

        mimetype = 'application/rdf+xml'
        self.request.response.setHeader('Content-Type', mimetype)
        self.request.response.setHeader('Content-Length', len(contents))
        return contents


class ExposureRdfFileRedirect(ExposureFileRedirect):
    target_view = 'pmr2_rdf'


pmr2/rdf/base.py
from cStringIO import StringIO
from lxml import etree
import rdflib

from pmr2.rdf.graph import PMR2Graph


class RdfXmlObject(object):
    """
    RDF/XML helper class.
    """

    _base_namespaces = {
        'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
    }

    format = 'xml'

    def __init__(self):
        self._purge()

    def _purge(self):
        self.graph = None
        self.dom = None
        self.subgraphIds = []

    @property
    def namespaces(self):
        return self._base_namespaces

    def parse(self, input):
        """\
        Parses a file.  Will overwrite all attributes.
        """

        self._purge()

        if not hasattr(input, 'read'):
            input = StringIO(input)

        self.graph = PMR2Graph()
        self.dom = etree.parse(input)
        self.subgraphIds = []  # maps line number in source to subgraph

        rdfnodes = []
        xpath_expr = [
            './/rdf:RDF',
            '/rdf:RDF',
        ]
        for expr in xpath_expr:
            rdfnodes.extend(self.dom.xpath(expr, namespaces=self.namespaces))

        for node in rdfnodes:
            s = StringIO(etree.tostring(node))
            s.seek(0)
            subgraph = self.graph.parse(s, format=self.format)
            self.subgraphIds.append((node.sourceline, subgraph.identifier))

    def mergeNs(self, otherNs):
        ns = {}
        ns.update(self.namespaces)
        ns.update(otherNs)
        return ns

    def query(self, q, initBindings={}, otherNs={}):
        ns = self.mergeNs(otherNs)
        return self.graph.query(q, initBindings=initBindings, initNs=ns)

    def queryKeys(self, node, keys, opt_keys=[], otherNs={}):
        """\
        This construct keys/value pair and dynamic simple query.
        """

        # XXX I want to be able to let all values be potentially 
        # optional, but doesn't seem like SPARQL is meant to do that.

        if isinstance(node, basestring):
            node = rdflib.URIRef(node)

        bindings = {
            rdflib.Variable('?node'): node,
        }

        base_q = 'SELECT %s WHERE \n{\n%s}'
        base_stmt = '    ?node %s %s .\n'
        opt_stmt  = '    OPTIONAL { ?node %s %s . }\n'
        frag_key = []
        frag_stmt = []
        for k in keys:
            ns, elem = k.split(':')
            q_key = '?' + elem
            frag_key.append(q_key)
            if k in opt_keys:
                frag_stmt.append(opt_stmt % (k, q_key))
            else:
                frag_stmt.append(base_stmt % (k, q_key))

        q = base_q % (' '.join(frag_key), ''.join(frag_stmt))

        ns = self.mergeNs(otherNs)
        self._lastq = q
        return self.graph.query(q, initBindings=bindings, initNs=ns)


class RdfXmlMetadata(RdfXmlObject):
    """\
    Some sample methods for queries.
    """

    _base_namespaces = {
        'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
        'pmr2': 'http://namespace.physiomeproject.org/pmr2#',
        'pmr2note': 'http://namespace.physiomeproject.org/pmr2/note#',
        'dc': 'http://purl.org/dc/elements/1.1/',
        'dcterms': 'http://purl.org/dc/terms/',
        'vCard': 'http://www.w3.org/2001/vcard-rdf/3.0#',
    }

    def queryEFNote(self, name):
        """\
        A rough way to get data that may be used to populate an exposure
        file note.
        """

        subject = self.namespaces['pmr2note'] + name
        bindings = {
            rdflib.Variable('?type'): rdflib.URIRef(subject),
        }
        q = """\
        SELECT ?key ?value WHERE {
            ?node pmr2:type ?type .
            ?node pmr2:fields [ ?li ?fnode ] .
            ?fnode pmr2:field [ pmr2:key ?key ] .
            ?fnode pmr2:field [ pmr2:value ?value ] .
        }
        """
        results = self.query(q, bindings, self.namespaces)
        return [(i[0].strip(), i[1].strip()) for i in results]

    def queryDC(self, subject):
        """\
        Returns some Dublin Core information from the target node.
        """

        keys = ['dc:title', 'dc:creator', 'dc:description']
        opt_keys = ['dc:creator', 'dc:description']
        result = self.queryKeys(subject, keys, opt_keys, self.namespaces)
        return [tuple(zip(keys, [j and j.strip() for j in i])) for i in result]
